app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -1,34 +1,70 @@
 package com.lightricks.feedexercise.ui.feed
 
+import android.util.Log
 import androidx.lifecycle.*
 import com.lightricks.feedexercise.data.FeedItem
 import com.lightricks.feedexercise.data.FeedRepository
 import com.lightricks.feedexercise.network.FeedApiService
 import com.lightricks.feedexercise.util.Event
 import java.lang.IllegalArgumentException
+import com.lightricks.feedexercise.R
 
 /**
  * This view model manages the data for [FeedFragment].
  */
 open class FeedViewModel(private val feedRepository: FeedRepository) : ViewModel() {
-    private val isLoading = MutableLiveData<Boolean>()
-    private val isEmpty = MutableLiveData<Boolean>()
-    private val feedItems = MediatorLiveData<List<FeedItem>>()
+    private val TAG = "FeedViewModel"
+
+    private val stateLiveData = MutableLiveData<State>()
     private val networkErrorEvent = MutableLiveData<Event<String>>()
 
-    fun getIsLoading(): LiveData<Boolean> = isLoading
-    fun getIsEmpty(): LiveData<Boolean> = isEmpty
-    fun getFeedItems(): LiveData<List<FeedItem>> = feedItems
+    fun getIsLoading(): LiveData<Boolean> = Transformations.map(stateLiveData) { it.isLoading }
+    fun getIsEmpty(): LiveData<Boolean> = Transformations.map(stateLiveData) { it.isFeedEmpty() }
+    fun getFeedItems(): LiveData<List<FeedItem>> =
+        Transformations.map(stateLiveData) { it.feedItems }
+
     fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
 
     init {
+        observeRepository()
         refresh()
     }
 
     fun refresh() {
-        //todo: fix the implementation
-        isLoading.value = false
-        isEmpty.value = true
+        setStartLoading()
+        feedRepository.fetchFeed()
+            .subscribe({ setStoppedLoading() },
+                { error -> handleNetworkError(error) })
+    }
+
+    private fun handleNetworkError(error: Throwable?) {
+        networkErrorEvent.postValue(Event(R.string.repository_network_error.toString()))
+        val message = error?.message
+        if (message != null)
+            Log.e(TAG, message)
```

Comment:
Add brackets even for condition we one statement. 

http://www.fredosaurus.com/notes-java/flow/if/30if-braces.html

Reply:
Also in this case I would always add a custom error message and pass the error to the logger. That way we always have a custom message. Also we always log the error we get. 

`Log.e(TAG, "Failed to fetch feed from network", error);`

Reply:
I decided to use the suggestion in the last comment, so it removed the need for a condition.